core/runtime/src/main/java/io/quarkus/runtime/ApplicationLifecycleManager.java

```diff
@@ -429,13 +429,15 @@ public void run() {
             } finally {
                 stateLock.unlock();
             }
-            if (currentApplication.isStarted()) {
+            //take a reliable reference before changing the application state:
+            final Application app = currentApplication;
+            if (app.isStarted()) {
                 // On CLI apps, SIGINT won't call io.quarkus.runtime.Application#stop(),
                 // making the awaitShutdown() below block the application termination process
                 // It should be a noop if called twice anyway
-                currentApplication.stop();
+                app.stop();
             }
-            currentApplication.awaitShutdown();
+            app.awaitShutdown();
```

Comment:
This is the line were it would NPE: the "stop" process invoked above would set the `currentApplication` field to `null`, asynchronously.
On my machine this would result in an NPE being thrown sometimes (not always).